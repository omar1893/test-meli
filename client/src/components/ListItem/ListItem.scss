@import "../../../src/index.scss";

.listItem {
    cursor: pointer;
    @include grid-basic(10);

    
    .listItem__image {
        width: 18rem;
        height: 18rem;
        border-radius: 0.4rem;
        margin-right: 1.6rem;
        grid-column-start: 1;

        @media screen and (max-width: 700px) {
            grid-column-start: 1;
            grid-column-end: 11;
            margin: 0 auto;
        }
    }
    
    .listItem__info-container {
        display: flex;
        grid-column-start: 3;
        grid-column-end: 9;
    }
    .listItem__info{
        .listItem__info--price {
            font-size: 2.4rem;
            margin-bottom: 3.2rem;
            @media screen and (max-width: 700px) {
                text-align: center;
            }
        }

        .listItem__info--title {
            font-size: 1.8rem;
            @media screen and (max-width: 700px) {
                text-align: center;
            }
        }

        .listItem__info--icon {
            margin-left: 1rem;
        }
    }

    .listItem__location {
        grid-column-start: 9;
        grid-column-end: 11;
        @media screen and (max-width: 700px) {
            grid-column-start: 1;
        }

        .listItem__location--city {
            font-size: 1.4rem;
            color: $dark-gray;

            @media screen and (max-width: 700px) {
                text-align: center;
            }
        }
    }
}