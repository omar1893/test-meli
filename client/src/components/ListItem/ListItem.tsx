import React from "react";
import "./ListItem.scss";
import FreeShipping from "../../assets/ic_shipping.png";
import { useNavigate } from "react-router-dom";
import { priceFormatter } from "../../utils/priceFormatter";
import { ListItemProps } from "../../interfaces";

const ListItem = ({ item }: ListItemProps) => {
  const navigate = useNavigate();

  const handleItemClick = () => {
    navigate(`/items/${item.id}`);
  };

  return (
    <div className="listItem" onClick={handleItemClick}>
      <img className="listItem__image" src={item.picture} alt="" />
      <div className="listItem__info-container">
        <div className="listItem__info">
          <p className="listItem__info--price">
            {priceFormatter(item.price.amount, item.price.currency)}
            {item.free_shipping && (
              <img className="listItem__info--icon" src={FreeShipping} alt="Free Shipping" />
            )}
          </p>
          <p className="listItem__info--title">{item.title}</p>
        </div>
      </div>
      <div className="listItem__location">
        <p className="listItem__location--city">{item.address}</p>
      </div>
    </div>
  );
};

export default ListItem;
