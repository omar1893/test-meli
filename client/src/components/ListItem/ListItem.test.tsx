import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import ListItem from "./ListItem";
import { useNavigate } from "react-router-dom";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: jest.fn(),
}));

describe("ListItem", () => {

  const author = {
    name: 'Omar',
    lastname: 'Lopez'
  }
  const item = {
    id: 1,
    picture: "item-picture.jpg",
    price: {
      amount: 100,
      currency: "USD",
      decimal: 0,
    },
    free_shipping: true,
    title: "Item Title",
    address: "Item Address",
    condition: 'new',
    sold_quantity: 100,
    description: ''
  };

  it("Should match snapshot", () => {
    const { container } = render(<ListItem item={item} author={author}/>);
    expect(container).toMatchSnapshot();
  });

  it("Should display item details correctly", () => {
    render(<ListItem item={item} author={author}/>);

    const priceElement = screen.getByText("100,00 US$");
    const titleElement = screen.getByText("Item Title");
    const addressElement = screen.getByText("Item Address");
    const shippingIconElement = screen.getByAltText("Free Shipping");

    expect(priceElement).toBeInTheDocument();
    expect(titleElement).toBeInTheDocument();
    expect(addressElement).toBeInTheDocument();
    expect(shippingIconElement).toBeInTheDocument();
  });

  it("Should navigate to item details on click", () => {
    const navigateMock = jest.fn();
    (useNavigate as jest.Mock).mockImplementation(() => navigateMock);

    const { container } = render(<ListItem item={item} author={author}/>);

    const listItemElement = container.querySelector(".listItem") as Element;
    fireEvent.click(listItemElement)

    expect(navigateMock).toHaveBeenCalledWith("/items/1");
  });
});
