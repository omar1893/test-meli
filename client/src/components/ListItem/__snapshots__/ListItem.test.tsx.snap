// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Should match snapshot 1`] = `
<div>
  <div
    class="listItem"
  >
    <img
      alt=""
      class="listItem__image"
      src="item-picture.jpg"
    />
    <div
      class="listItem__info-container"
    >
      <div
        class="listItem__info"
      >
        <p
          class="listItem__info--price"
        >
          $100.00
          <img
            alt="Free Shipping"
            class="listItem__info--icon"
            src="ic_shipping.png"
          />
        </p>
        <p
          class="listItem__info--title"
        >
          Item Title
        </p>
      </div>
    </div>
    <div
      class="listItem__location"
    >
      <p
        class="listItem__location--city"
      >
        Item Address
      </p>
    </div>
  </div>
</div>
`;
